### DISCUSSION TITLE
Merge branches keeps strings with same identifier

### DISCUSSION THREAD

**Discussion by jimmysafe** [crowdin/crowdin-cli#787] [Q&A]
**Describe the bug**

We are using String Based Projects. 
We have a `main` branch which contains all the strings available in production.
Then for every feature we develop we create a separate Crowdin branch containing all the strings for that specific feature **but it could also include strings already present in main which have been modified.**

To explain this lets say we have a `main` and a `feat` crowdin branches.

we have a string in `main`:
```
identifier:  login.hello
text: Hello
```

We also have a string in `feat` (this is the string that counts, and that we have specifically modified)
```
identifier:  login.hello
text: Hello Bud!
```

**Now, whenever we merge the `feat` branch into `main` branch we see that both strings with same identifier have been added to main!**
So we get both strings `Hello Bud!` and `Hello` with same identifier `login.hello`.

Is this an expected behaviour? 

**We would like to keep only the new string contained in `feat` branch when we merge into `main`. We don't expect to see the old string.**

**To Reproduce**
1. Create 2 crowdin branches
2. Add 2 strings respectively with same identifier but different text.
3. Merge branch 1 in branch 2
4. See that in branch 1 there are 2 strings with same identifier and different text.

**Expected behavior**
When merging branch 2 into branch 1 all OLD strings with duplicate identifier should be removed in favour of new one.

**Screenshots**
![Screenshot 2024-05-10 at 20 53 35](https://github.com/crowdin/crowdin-cli/assets/12184812/cd7dbb54-4668-4e9d-8b63-f251ed4198e7)

**Comment by andrii-bodnar** [comment]
Hi @jimmysafe, did you clone the `feat` branch from the `main` branch?

The expected behavior is that when merging the `feat` branch into the `main` branch, all old strings with duplicate identifiers should be removed in favor of the new ones.

Here's a workflow to clarify:

1. If a user clones from the branch and changes the text in the cloned branch, only one string should remain after the merge.
2. If a user clones from the branch, adds a string to the main, and then to the feat branch, there will be two strings.
3. At the time of cloning, there is a mapping of strings between the main and feat. If the strings were not in the main at that time, there wouldn't be a mapping, and we consider it a new string, even if it has the same key or text.

The workflow should be approximately as follows:

- User pushes sources to the main.
- Clone branch feature from the main.
- User adds/edits strings in the feat.
- Then, merge, and depending on what was done in both branches, we either merge changes into the master (if new strings were added or only edited in one of the branches) or resolve conflicts (if the same string was edited in both master and feature branches).

**Comment by jimmysafe** [comment]
Hi @andrii-bodnar thanks for the clarification.

So basically you have to clone the `main` branch into `feat` in order to make this work, that makes sense.

The reason why we did not clone the main branch in the first place is because the CLI and the APIv2 does not have an endpoint to check che conflicts when merging. 

I can see this endpoint available on your frontend, but not in the API:
`https://api.crowdin.com/api/v2/projects/PROJECTID/branches/BRANCHID/merges/MERGEID/conflicts`

Is there any chance to get this endpoint added to the API?

It would solve so many issues for us, as we are trying to automate a custom process with the crowdin api.

**Comment by andrii-bodnar** [comment]
@jimmysafe, yes, I have already passed it on to the team to consider making this endpoint public.

By the way, we will also be adding the `crowdin branch clone`, and `crowdin branch merge` commands to the CLI in the near future.

**Comment by jimmysafe** [comment]
That is amazing! thanks!

With the conflicts endpoint it would also be great to have and endpoint to confirm merge conflicts selection as you do on your frontend, so the whole picture is complete :) 

Btw thanks for the support, really appreciate <3

**Comment by jimmysafe** [comment]
@andrii-bodnar  hi! is there any update regarding the confirm merge endpoint? any ETA?
